### ISSUE TITLE
Cannot define a field with arbitrary structure

### ISSUE THREAD

**Issue by nicolaferraro** [kubernetes-sigs/controller-runtime#628]
I'm trying to define a controller for a type with some unknown parts in the tree that can contain arbitrary structures.

E.g.

```
apiVersion: my.dev/v1alpha1
kind: TypeWithSemistructuredData
metadata:
  name: thename
spec:
  replicas: 1
  myUnknownField:
    structure:
      ofThis:
        - isUnknown: true
      butStill:
        - yaml
        - json
        - compatible
```

Problem is how to map `myUnknownField` into an actual struct field. I've tried:

- `map[interface{}]interface{}`: controller runtime fails because cannot map `interface{}`
- `runtime.Unknown`: it's always nil
- `runtime.RawExtension`: ... well, it's a snippet, not a `runtime.Object`

Is there a clean way to define such fields that can contain objects with arbitrary structure?

**Comment by DirectXMan12** [comment]
Are the structures always maps?  `map[string]interface{}` should work from a go perspective

**Comment by nicolaferraro** [comment]
Yeah, I think I've found the solution looking at how `Unstructured` is defined in Kube APIs. Closing this with the solution for (my :D) future memory.

It's possible to define the unstructured part as:

```
type MyType map[string]interface{}

func (in *MyType) DeepCopy() *MyType {
	if in == nil {
		return nil
	}
	out := MyType(runtime.DeepCopyJSON(*in))
	return &out
}
```

Using this definition, including the DeepCopy method, then deepcopy-gen does generate the other required methods (DeepCopyInto, ...) based on it. Mapping at JSON/YAML level works.
Without explicit definition of DeepCopy, then deepcopy-gen seems to go into trouble.

The only difference between `map[string]interface{}` vs `map[interface{}]interface{}` is that the first level needs to be a JSON/Yaml object and not optionally an array, but I can live with it for my particular case.
